javascript/sbp/msg.js

```diff
@@ -0,0 +1,210 @@
+                 0x9188,0x81a9,0xb1ca,0xa1eb,0xd10c,0xc12d,0xf14e,0xe16f,
+                 0x1080,0x00a1,0x30c2,0x20e3,0x5004,0x4025,0x7046,0x6067,
+                 0x83b9,0x9398,0xa3fb,0xb3da,0xc33d,0xd31c,0xe37f,0xf35e,
+                 0x02b1,0x1290,0x22f3,0x32d2,0x4235,0x5214,0x6277,0x7256,
+                 0xb5ea,0xa5cb,0x95a8,0x8589,0xf56e,0xe54f,0xd52c,0xc50d,
+                 0x34e2,0x24c3,0x14a0,0x0481,0x7466,0x6447,0x5424,0x4405,
+                 0xa7db,0xb7fa,0x8799,0x97b8,0xe75f,0xf77e,0xc71d,0xd73c,
+                 0x26d3,0x36f2,0x0691,0x16b0,0x6657,0x7676,0x4615,0x5634,
+                 0xd94c,0xc96d,0xf90e,0xe92f,0x99c8,0x89e9,0xb98a,0xa9ab,
+                 0x5844,0x4865,0x7806,0x6827,0x18c0,0x08e1,0x3882,0x28a3,
+                 0xcb7d,0xdb5c,0xeb3f,0xfb1e,0x8bf9,0x9bd8,0xabbb,0xbb9a,
+                 0x4a75,0x5a54,0x6a37,0x7a16,0x0af1,0x1ad0,0x2ab3,0x3a92,
+                 0xfd2e,0xed0f,0xdd6c,0xcd4d,0xbdaa,0xad8b,0x9de8,0x8dc9,
+                 0x7c26,0x6c07,0x5c64,0x4c45,0x3ca2,0x2c83,0x1ce0,0x0cc1,
+                 0xef1f,0xff3e,0xcf5d,0xdf7c,0xaf9b,0xbfba,0x8fd9,0x9ff8,
+                 0x6e17,0x7e36,0x4e55,0x5e74,0x2e93,0x3eb2,0x0ed1,0x1ef0]
+
+// TODO: CRC implementation and verify CRC
+
+var mergeDict = function (dest, src) {
+  for (var prop in src) {
+    if (src.hasOwnProperty(prop)) {
+      dest[prop] = src[prop];
+    }
+  }
+  return dest;
+};
+
+function BufferTooShortError (message) {
+  this.name = 'BufferTooShortError';
+  this.message = message;
+  this.stack = (new Error()).stack;
+}
+BufferTooShortError.prototype = Object.create(Error.prototype);
+BufferTooShortError.prototype.constructor = BufferTooShortError;
+
+function BufferCorruptError (message) {
+  this.name = 'BufferCorruptError';
+  this.message = message;
+  this.stack = (new Error()).stack;
+}
+BufferCorruptError.prototype = Object.create(Error.prototype);
+BufferCorruptError.prototype.constructor = BufferCorruptError;
+
+var packages = ["acquisition", "bootload", "ext_events", "file_io", "flash", "logging", "navigation", "observation", "piksi", "settings", "system", "tracking"];
+
+var sbpTable = packages.map(function (pkg) {
+  return require(path.resolve(__dirname, "./" + pkg + ".js"));
+}).reduce(function (prev, curr) {
+  return mergeDict(prev, curr);
+}, {});
+
+var parser = new Parser()
+  .endianess('little')
+  .uint8('preamble')
+  .uint16('msg_type')
+  .uint16('sender')
+  .uint8('length')
+  .buffer('payload', { length: 'length' })
+  .uint16('crc');
+
+/**
+ * CRC16 implementation according to CCITT standards.
+ * Based on msg.py implementation.
+ */
+function crc16 (buf, crc) {
+  crc = crc || 0;
+
+  for (var i = 0; i < buf.length; i++) {
+    var ch = buf[i];
+    crc = ((crc<<8)&0xFFFF) ^ (crc16tab[((crc>>8)&0xFF) ^ (ch&0xFF)]);
+    crc &= 0xFFFF;
+  }
+
+  return crc;
+}
+
+module.exports = {
+  decode: function decode (msg) {
+    var sbp = parser.parse(msg);
+
+    return new sbpTable[sbp['msg_type']](sbp);
```

Comment:
Hey, so what happens here if the dispatch fails because the `msg_type` is missing? I'm assuming you get some kind of dictionary-related exception/error. In the past, we've caught these, logged a warning, and passed the raw SBP message through.

https://github.com/swift-nav/libsbp/blob/master/python/sbp/table.py#L68